File: src/backend/cs/StudyRoomSystem/StudyRoomSystem.Core/Structs/Entity/Violation.cs
using System.Text.Json.Serialization;

namespace StudyRoomSystem.Core.Structs.Entity;

[JsonConverter(typeof(JsonStringEnumConverter<ViolationStateEnum>))]
public enum ViolationStateEnum
{
    Violation
}

[JsonConverter(typeof(JsonStringEnumConverter<ViolationTypeEnum>))]
public enum ViolationTypeEnum
{
    超时,
    强制取消,
    管理员
}

/// <summary>
/// 违规记录
/// </summary>
public record Violation
{
    public Guid Id { get; set; } = Ulid.NewUlid().ToGuid();

    public Guid? BookingId { get; set; }
    public required Guid UserId { get; set; }
    public DateTime CreateTime { get; set; } = DateTime.UtcNow;
    public required ViolationStateEnum State { get; set; }
    public required ViolationTypeEnum Type { get; set; }
    public required string Content { get; set; }

    public virtual User User { get; set; } = null!;
    public virtual Booking? Booking { get; set; }
}

File: src/frontend/web/study-room-system/src/lib/types/Violation.ts
import type { Booking } from "./Booking";
import type { User } from "./User";

export type ViolationState = "Violation";
export type ViolationType = "超时" | "强制取消" | "管理员";

export interface Violation {
  id: string;
  bookingId: string | null;
  userId: string;
  createTime: string;
  state: ViolationState;
  type: ViolationType;
  content: string;
  user: User;
  booking: Booking | null;
}

export interface ViolationPage {
  total: number;
  page: number;
  pageSize: number;
  items: Violation[];
}

export interface ViolationCreateDto { 
  userId: string; 
  bookingId?: string | null; 
  type: "超时" | "强制取消" | "管理员"; 
  content: string; 
} 

export interface ViolationUpdateDto { 
  id: string; 
  type?: "超时" | "强制取消" | "管理员" | null; 
  content?: string | null; 
}

export function localizeViolationType(type: ViolationType): string {
  switch (type) {
    case "超时":
      return "超时";
    case "强制取消":
      return "强制取消";
    case "管理员":
      return "管理员处理";
    default:
      return type; // 兜底
  }
}
